crates/cargo-contract/src/cmd/extrinsics/balance.rs

```diff
@@ -0,0 +1,670 @@
+    ///        symbol: String::from("DOT"),
+    /// };
+    /// let sample_den_balance = BalanceVariant::Denominated(DenominatedBalance {
+    ///     value: Decimal::new(5005, 1),
+    ///     unit: UnitPrefix::Mega,
+    ///     symbol: String::from("DOT")
+    /// });
+    /// let balance: Balance = 5_005_000_000_000_000_000;
+    /// let den_balance = BalanceVariant::from(balance, Some(tm));
+    /// assert_eq!(Ok(sample_den_balance), Ok(den_balance));
+    /// ```
+    pub fn from<T: Into<u128>>(
+        value: T,
+        token_metadata: Option<&TokenMetadata>,
+    ) -> Result<Self> {
+        let n: u128 = value.into();
+
+        if let Some(token_metadata) = token_metadata {
+            if n == 0 {
+                return Ok(BalanceVariant::Denominated(DenominatedBalance {
+                    value: Decimal::ZERO,
+                    unit: UnitPrefix::One,
+                    symbol: token_metadata.symbol.clone(),
+                }))
+            }
+
+            let number_of_digits = n.to_string().len();
+
+            let giga_units_zeros = token_metadata.token_decimals + 9;
+            let mega_units_zeros = token_metadata.token_decimals + 6;
+            let kilo_units_zeros = token_metadata.token_decimals + 3;
+            let one_unit_zeros = token_metadata.token_decimals;
+            let milli_units_zeros = token_metadata.token_decimals.checked_sub(3);
+            let micro_units_zeros = token_metadata.token_decimals.checked_sub(6);
+            let nano_units_zeros = token_metadata.token_decimals.checked_sub(9);
+
+            let multiple: Decimal;
+            let unit: UnitPrefix;
+            if (giga_units_zeros + 1..).contains(&number_of_digits) {
+                multiple = Decimal::from_str_exact(&format!(
+                    "1{}",
+                    "0".repeat(giga_units_zeros)
+                ))?;
+                unit = UnitPrefix::Giga;
+            } else if (mega_units_zeros + 1..=giga_units_zeros)
+                .contains(&number_of_digits)
+            {
+                multiple = Decimal::from_str_exact(&format!(
+                    "1{}",
+                    "0".repeat(mega_units_zeros)
+                ))?;
+                unit = UnitPrefix::Mega;
+            } else if (kilo_units_zeros + 1..=mega_units_zeros)
+                .contains(&number_of_digits)
+            {
+                multiple = Decimal::from_str_exact(&format!(
+                    "1{}",
+                    "0".repeat(kilo_units_zeros)
+                ))?;
+                unit = UnitPrefix::Kilo;
+            } else if (one_unit_zeros + 1..=kilo_units_zeros).contains(&number_of_digits)
+            {
+                multiple =
+                    Decimal::from_str_exact(&format!("1{}", "0".repeat(one_unit_zeros)))?;
+                unit = UnitPrefix::One;
+            } else if milli_units_zeros.is_some()
+                && (milli_units_zeros.unwrap() + 1..=one_unit_zeros)
```

Comment:
Instead of unwrapping at the usage site, maybe use an `expect` at the declaration site that we don't expect `token_deciamls` < 3. Although maybe we should  allow it just any high precision values will not be displayed as denominated